### PR TITLE
[SPARK-43300][CORE] NonFateSharingCache wrapper for Guava Cache

### PR DESCRIPTION
**Pull request by liuzqt**
### What changes were proposed in this pull request?

Create `NonFateSharingCache` to wrap around Guava cache with a KeyLock to synchronize all requests with the same key, so they will run individually and fail as if they come one at a time.

Wrap cache in `CodeGenerator` with `NonFateSharingCache` to protect it from unexpected cascade failure due to cancellation from irrelevant queries that loading the same key. Feel free to use this in other places where we used Guava cache and don't want fate-sharing behavior.

Also, instead of implementing Guava Cache and LoadingCache interface, I define a subset of it so that we can control at compile time what cache operations are allowed and make sure all cache loading action go through our narrow waist code path with key lock. Feel free to add new APIs when needed.

### Why are the changes needed?

Guava cache is widely used in spark, however, it suffers from fate-sharing behavior: If there are multiple requests trying to access the same key in the cache at the same time when the key is not in the cache, Guava cache will block all requests and create the object only once. If the creation fails, all requests will fail immediately without retry. So we might see task failure due to irrelevant failure in other queries due to fate sharing.

This fate sharing behavior leads to unexpected results in some situation(for example, in code gen).

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?
UT

### CODE REVIEW

File: core/src/test/scala/org/apache/spark/util/NonFateSharingCacheSuite.scala
Comment (ryan-johnson-databricks) [inline core/src/test/scala/org/apache/spark/util/NonFateSharingCacheSuite.scala:123]: nit: indent these 4 spaces?

File: core/src/test/scala/org/apache/spark/util/NonFateSharingCacheSuite.scala
Comment (ryan-johnson-databricks) [inline core/src/test/scala/org/apache/spark/util/NonFateSharingCacheSuite.scala:42]: nit: use Option instead of null check?
```scala
if (Option(THREAD2_HOLDER.get()).exists(_.getState.equals(Thread.State.WAITING))) {
```

File: core/src/test/scala/org/apache/spark/util/NonFateSharingCacheSuite.scala
Comment (ryan-johnson-databricks) [inline core/src/test/scala/org/apache/spark/util/NonFateSharingCacheSuite.scala:100]: Can we use more descriptive names for test cases? 
Ideally it explains what behavior the test actually verifies.

File: core/src/main/scala/org/apache/spark/util/NonFateSharingCache.scala
Comment (ryan-johnson-databricks) [inline core/src/main/scala/org/apache/spark/util/NonFateSharingCache.scala:37]: I don't think the "feel free to..." comment is especially helpful... can probably just remove it.

File: core/src/main/scala/org/apache/spark/util/NonFateSharingCache.scala
Comment (ryan-johnson-databricks) [inline core/src/main/scala/org/apache/spark/util/NonFateSharingCache.scala:40]: Worth a quick doc comment to explain that this method will return a `NonFateSharingLoadingCache` if the user happens to pass a `LoadingCache`, as a courtesy to the user?

File: core/src/main/scala/org/apache/spark/util/NonFateSharingCache.scala
Comment (ryan-johnson-databricks) [inline core/src/main/scala/org/apache/spark/util/NonFateSharingCache.scala:35]:
```suggestion
 * Instead of implementing Guava Cache and LoadingCache interface, we expose a subset of APIs
```

File: core/src/main/scala/org/apache/spark/util/NonFateSharingCache.scala
Comment (ryan-johnson-databricks) [reply core/src/main/scala/org/apache/spark/util/NonFateSharingCache.scala:37]: Meanwhile, it would probably be helpful to explain _why_ we don't/can't/won't expose the full Guava cache API?

File: core/src/test/scala/org/apache/spark/util/NonFateSharingCacheSuite.scala
Comment (liuzqt) [reply core/src/test/scala/org/apache/spark/util/NonFateSharingCacheSuite.scala]: I was hitting a overload resolution issue:
```
ambiguous reference to overloaded definition,
[error] both method submit in class AbstractExecutorService of type [T](x$1: java.util.concurrent.Callable[T])java.util.concurrent.Future[T]
[error] and  method submit in class AbstractExecutorService of type (x$1: Runnable)java.util.concurrent.Future[_]

```
But I can do 
```
val r1: Runnable = () => thread1Task()
val f1 = executor1.submit(r1)
```
instead

Comment (ryan-johnson-databricks) [reply core/src/test/scala/org/apache/spark/util/NonFateSharingCacheSuite.scala]: Oh that's right... spark added a bunch of "helpful" overloads that tried to compensate for older JVM... but on newer JVM with runnable interface support the overload gets in the way.

File: core/src/main/scala/org/apache/spark/util/NonFateSharingCache.scala
Comment (ryan-johnson-databricks) [inline core/src/main/scala/org/apache/spark/util/NonFateSharingCache.scala:30]: Maybe add a final connection like:
> Given that spark can cancel tasks at arbitrary times for many different reasons, fate sharing means that a task which gets canceled while populating a cache entry can cause spurious failures in tasks from unrelated jobs -- even though those tasks would have successfully populated the cache if they had been allowed to try.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala
Comment (ryan-johnson-databricks) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala:1583]: If we move some of this to the NonFateSharingCache doc comment, it might suffice to say something simpler here, like:
> Codegen can be slow. Use a non fate sharing cache in case a query gets canceled during code gen while other queries wait on the same code, so that those other queries don't get wrongly aborted. See [[NonFateSharingCache]] for more details.

**Review by ryan-johnson-databricks — APPROVED** [review]
LGTM, just a couple comment wording nits

**Comment by ryan-johnson-databricks** [comment]
PR description seems a bit vague, given that actual code comments give pretty clear reasoning for why this change is needed?

> Wrap cache in `CodeGenerator` as an example. Feel free to use this in other places where we used Guava cache and don't want fate-sharing behavior.

It's not just an "example" -- the code comments detail a specific bad behavior we're trying to avoid with this change.

> This fate sharing behavior might lead to unexpected results in some situation.

Again, we can be specific -- we know it _does_ lead to unexpected results (query canceled during codegen)

**Comment by liuzqt** [comment]
Hi @ryan-johnson-databricks would you mind triggering the merge for this PR?

**Comment by ryan-johnson-databricks** [comment]
> Hi @ryan-johnson-databricks would you mind triggering the merge for this PR?

Sorry, I'm not a spark committer.

File: core/src/main/scala/org/apache/spark/util/NonFateSharingCache.scala
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/util/NonFateSharingCache.scala]: Can we make this `private[spark]` so that user code can't take a dependency on it?

File: core/src/main/scala/org/apache/spark/util/NonFateSharingCache.scala
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/util/NonFateSharingCache.scala]: Similarly, `private[spark]` here as well?

File: core/src/main/scala/org/apache/spark/util/NonFateSharingCache.scala
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/util/NonFateSharingCache.scala]: Similarly, `private[spark]` here as well?

**Review by JoshRosen — APPROVED** [review]
LGTM (pending a class visibility change to add in `private[spark]` in a few places).

**Edit:** to provide some additional context on motivation for this change: this addresses a "task cancellation triggers failures of other tasks" problem which can be triggered when redundant speculative tasks are canceled. The re-usable `NonFateSharingCache` wrapper will make it easy to apply the same fix in other contexts where fate-sharing can occur through Guava caches.

**Comment by liuzqt** [comment]
Hi @JoshRosen would you mind merging this PR? Or maybe @cloud-fan could help?

**Comment by JoshRosen** [comment]
I've merged this to master (Spark 3.5.0). Thanks @liuzqt!

**Comment by LuciferYang** [comment]
@liuzqt @JoshRosen Report an [issue]([SPARK-44064](https://issues.apache.org/jira/browse/SPARK-44064)) about maven test related to this pr.

The `CodeGenerator` in the catalyst module uses the `NonFateSharingCache` in the core module, and the input parameter of `NonFateSharingCache#apply` is `com.google.common.cache.Cache`.

The catalyst module may use shaded spark-core jar when we do maven test, the Guava related classes in the core module will be relocated from `com.google.common` to `org.sparkproject.guava`, so the input parameter of `NonFateSharingCache#apply` will change to `org.sparkproject.guava.cache.Cache`, but the catalyst module has not been shaded yet when do maven test, so `CodeGenerator` will still use `com.google.common.cache.Cache` to call the `NonFateSharingCache#apply` function, this will result in a mismatch of input types when do maven test and maven test will failed.

The steps to reproduce are as follows:

```
 build/mvn clean install -DskipTests -pl sql/catalyst -am
 build/mvn test -pl sql/catalyst 
```

the catalyst module maven test aborted as follows:

```
ProductAggSuite:
*** RUN ABORTED ***
  java.lang.NoClassDefFoundError: Could not initialize class org.apache.spark.sql.catalyst.expressions.codegen.CodeGenerator$
  at org.apache.spark.sql.catalyst.expressions.codegen.JavaCode$.variable(javaCode.scala:64)
  at org.apache.spark.sql.catalyst.expressions.codegen.JavaCode$.isNullVariable(javaCode.scala:77)
  at org.apache.spark.sql.catalyst.expressions.Expression.$anonfun$genCode$3(Expression.scala:200)
  at scala.Option.getOrElse(Option.scala:189)
  at org.apache.spark.sql.catalyst.expressions.Expression.genCode(Expression.scala:196)
  at org.apache.spark.sql.catalyst.expressions.codegen.GenerateSafeProjection$.$anonfun$create$1(GenerateSafeProjection.scala:156)
  at scala.collection.immutable.List.map(List.scala:293)
  at org.apache.spark.sql.catalyst.expressions.codegen.GenerateSafeProjection$.create(GenerateSafeProjection.scala:153)
  at org.apache.spark.sql.catalyst.expressions.codegen.GenerateSafeProjection$.create(GenerateSafeProjection.scala:39)
  at org.apache.spark.sql.catalyst.expressions.codegen.CodeGenerator.generate(CodeGenerator.scala:1369) 
```

I revert this patch locally, and the above maven test can pass.

also cc @HyukjinKwon to know

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala:1584]: In that case, I guess we should either move this to core module, or `NonFateSharingCache` to catalyst module?

Do you have other better ideas, @LuciferYang ?

**Pull request closed without merging**
